src/btree/bt_sync.c

```diff
@@ -109,44 +109,53 @@ __sync_ref_is_obsolete(WT_SESSION_IMPL *session, WT_REF *ref)
 {
     WT_ADDR *addr;
     WT_CELL_UNPACK vpack;
+    WT_PAGE_MODIFY *mod;
 
     /* Ignore root pages as they can never be deleted. */
     if (__wt_ref_is_root(ref))
         return (false);
 
-    /* Guard against marking an in-memory page to be deleted. */
-    if (ref->state != WT_REF_DISK)
-        return (false);
-
     /* Ignore internal pages, these are taken care of during reconciliation. */
     if (!__wt_ref_is_leaf(session, ref))
         return (false);
 
+    /* In-memory ref address must be from an off page address. */
+    mod = (ref->page != NULL) ? (ref->page->modify) : NULL;
+    if (mod != NULL && mod->rec_result == WT_PM_REC_REPLACE)
+        return (__wt_txn_visible_all(
+          session, mod->mod_replace.newest_stop_txn, mod->mod_replace.newest_stop_ts));
+
+    if (mod != NULL && mod->rec_result == WT_PM_REC_MULTIBLOCK)
+        return (__wt_txn_visible_all(
+          session, mod->mod_multi_newest_stop_txn, mod->mod_multi_newest_stop_ts));
+
     WT_ORDERED_READ(addr, ref->addr);
+    WT_ASSERT(session, addr != NULL);
```

Comment:
I don't believe you need an ordered read here, can you please write down what made you think one was necessary?

There's not much point in asserting a value isn't NULL, is there? We're about to drop core if it's NULL anyway.

Reply:
I don't find a scenario, so removed it. Added a condition `addr != NULL` to find out scenarios where the page is not reconciled yet and not disk address.